cpp/src/Ice/Instance.cpp

```diff
@@ -1734,6 +1734,36 @@ IceInternal::Instance::destroy()
     }
 }
 
+void
+IceInternal::Instance::destroyAsync(function<void()> completed) noexcept
+{
+    assert(completed); // the caller (Communicator) makes sure completed is callable.
+
+    bool executeCallback = false;
+    {
+        lock_guard lock(_mutex);
+        if (_state == StateDestroyed)
+        {
+            executeCallback = true; // execute outside the lock
+        }
+        else
+        {
+            // Start a background thread that calls destroy() and then executes the callback.
```

Comment:
I don't see how we could use a joinable thread here. Who would join it?

Reply:
Can't we join it in the destructor?

Reply:
The thread "body" keeps a refcount on Instance, and I assume it's kept while the thread exists. It's not clear to me how this join could work.

Reply:
> it's desirable it has a refcount on the instance.

If the instance owns the thread, the thread doesn't need to keep a refcount.

Reply:
You're right, I've updated the PR to join the threads in the destructor.